### PR TITLE
port Transformer

### PR DESCRIPTION
**Pull request by cgarciae**
# Description
* Port `Transformer` to NNX and creates the `transformer_to_linen` wrapper.

# Tests
* Llama2-7B gives same perf with `enable_nnx` on/off:
```
python3 -m MaxText.train MaxText/configs/base.yml \
    run_name=<run_name> \
    base_output_directory=gs://<gcs_bucket> \
    dataset_type=synthetic \
    steps=10 \
    model_name=llama2-7b
```
* Llama2-7B gives decode output with `enable_nnx` on/off:
```
python3 -m MaxText.decode MaxText/configs/base.yml \
    model_name=llama2-7b \
    tokenizer_path=assets/tokenizer_llama3.tiktoken \
    tokenizer_type=tiktoken \
    scan_layers=false \
    per_device_batch_size=1 \
    ici_fsdp_parallelism=1 \
    ici_autoregressive_parallelism=-1 \
    max_prefill_predict_length=128 \
    max_target_length=256 \
    prompt="I love to" \
    attention=dot_product
```
* Llama3.3-70B Maxengine/Jetstream gives same accuracy for `enable_nnx` on/off: https://diff.googleplex.com/#key=O4rlQHQDz8Cg

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -1098,10 +1099,7 @@ def get_kv_cache_annotations(model, config, rng, mesh, page_state: Optional[Page
   """Get a shaped abstraction of the state (including optimizer)"""
 
   def init_kv_cache(model, config):
-    input_shape = (
-        config.global_batch_size_to_load,
-        1,
-    )
+    input_shape = (config.micro_batch_size_to_train_on, 1)
```
Comment (bvandermoon) [inline MaxText/maxtext_utils.py:1107]: This should be changed in `get_prefill_kv_cache_annotations also, right?

**Review by bvandermoon — COMMENTED** [review]
Discussed the inference memory issues more with the team this morning. I am looking into adding a config flag to enable/disable NNX for the remaining modules. We can use that flag for this and https://github.com/AI-Hypercomputer/maxtext/pull/2178. That way we can still merge to main. Hoping it is fairly straightforward, taking a closer look now

File: MaxText/maxtext_utils.py (hunk 1 above)
Comment (cgarciae) [reply MaxText/maxtext_utils.py:1107]: yeah that looks about right

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -289,6 +289,29 @@ def dense_general(
...
+    rng_collection: str = 'dropout',
+    rngs: nnx.Rngs| None = None,
+  ):
+    self.rate = rate
+    self.broadcast_dims = broadcast_dims
+    self.deterministic = deterministic
+    self.rng_collection = rng_collection
+
+    if isinstance(rngs, nnx.Rngs):
+      self.rngs = rngs.fork()
```
Comment (bvandermoon) [inline MaxText/layers/linears.py]: Could you add the fix you previously mentioned for old Flax versions here? I see an attribute error on Flax 0.10.6 from this line

```
 self.rngs = rngs.fork() if hasattr(type(rngs), 'fork') else rngs
```

File: MaxText/maxtext_utils.py (hunk 1 above)
Comment (bvandermoon) [reply MaxText/maxtext_utils.py:1107]: Reopening this, could you please also change this above?

File: MaxText/layers/decoders.py [hunk 3]
```diff
@@ -587,11 +587,11 @@ def _apply_output_head(self, y, deterministic, model_mode):
   @nn.compact
   def __call__(
       self,
+      shared_embedding: nn.Module | nnx.Module,
```
Comment (bvandermoon) [inline MaxText/layers/decoders.py:590]: Why do we need to pass this `shared_embedding` around now?

File: MaxText/maxengine.py [hunk 4]
```diff
@@ -1416,6 +1416,7 @@ def init(abstract_params, page_state):
           mutable=["cache"],
           page_state=page_state,
           slot=0,
+          # nnx_method="no_op",
```
Comment (bvandermoon) [inline MaxText/maxengine.py]: Nit: please remove commented code

Comment (bvandermoon) [reply MaxText/maxengine.py]: Or is it needed here?

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -796,3 +796,6 @@ projector_dropout_for_vit: 0.0
 # Subslice shape in the form of "x,y,z" when using pathways (single controller). 
 # Example: "8,8" to use a 8x8 subgrid (64 chips) of a full pod (16x16) of trillium.
 subslice_shape: ""
+
+# NNX
+enable_nnx: false
```
Comment (A9isha) [inline MaxText/configs/base.yml]: nit: please add the new line

File: MaxText/layers/decoders.py (hunk 3 above)
Comment (A9isha) [reply MaxText/layers/decoders.py:590]: `class Decoder` now becomes NNX, and shouldn't have a Linen attribute `shared_embedding`.  I wonder why we cannot use the NNX version of [embeddimgs](https://github.com/AI-Hypercomputer/maxtext/blob/a0a77e1d6adda33f49c8e14b3163ffe818300c56/MaxText/layers/models.py#L60). Does the bridge not support such nested NNX modules?

Comment (A9isha) [reply MaxText/layers/decoders.py:590]: My followup qs is also why can't we pass around `model_mode` anymore and need to save it as an attribute?

File: MaxText/layers/decoders.py [hunk 6]
```diff
@@ -751,7 +749,7 @@ def __call__(
                   decoder_segment_ids,
                   decoder_positions,
                   deterministic,
-                  model_mode,
+                  self.model_mode,
```
Comment (A9isha) [inline MaxText/layers/decoders.py:752]: sorry for the rookie question - why are we able to pass `model_mode` here but not above?

File: MaxText/layers/models.py [hunk 7]
```diff
@@ -52,6 +55,22 @@ class Transformer(nn.Module):
   model_mode: str = MODEL_MODE_TRAIN  # May be different than the model_mode passed to __call__
   # pylint: enable=attribute-defined-outside-init
 
+  def init(self, *args, model_mode: str = MODEL_MODE_TRAIN, **kwargs):
+    """Initializes the model."""
+    if model_mode != self.model_mode:
+      module = self.clone(model_mode=model_mode)
```
Comment (A9isha) [inline MaxText/layers/models.py]: I don't understand this :(

File: MaxText/layers/models.py [hunk 8]
```diff
@@ -161,12 +180,213 @@ def __call__(
...
+      true_length: int | None = None,
+      slot: int | None = None,
+      page_state: page_manager.PageState | None = None,
+      decoder_target_tokens: jax.Array | None = None,
+      decoder_target_mask: jax.Array | None = None,
+  ):
+    """Applies Transformer decoder-branch on encoded-input and target.
+
+    Args:
+      true_length: (Optional) Prompt length before padding
```
Comment (A9isha) [inline MaxText/layers/models.py:318]: wondering why only 2 of the args are described here...

File: MaxText/tests/model_test.py [hunk 9]
```diff
@@ -166,10 +167,8 @@ def test_train_vs_prefill_and_autoregress(self):
         mutable=["cache"],
     )
 
-    self.assertTrue(
-        jax.numpy.allclose(
-            full_train_logits[:, :PREFILL_RANGE, :], partial_prefill_logits, rtol=1e-01, atol=1e-01, equal_nan=False
-        )
+    np.testing.assert_allclose(
```
Comment (A9isha) [inline MaxText/tests/model_test.py:170]: curious why assertTrue doesn't work?

File: MaxText/train_utils.py [hunk 10]
```diff
@@ -35,10 +38,28 @@
...
+    model_mode: str = MODEL_MODE_TRAIN,
+) -> nn.Module: ...
+@overload
+def from_pretrained(
+    config: pyconfig.HyperParameters,
+    devices: Sequence[jax.Device] | None = None,
+    *,
+    model_mode: str = MODEL_MODE_TRAIN,
+    rngs: nnx.Rngs,
+) -> models.Transformer: ...
```
Comment (A9isha) [inline MaxText/train_utils.py:55]: why do we need two overridden methods? is it only to have non-null rngs? then should return type of the second one skip models.Transformer?

**Review by A9isha — COMMENTED** [review]
Hats off to you Cristian for making this work!!

File: MaxText/layers/decoders.py (hunk 3 above)
Comment (cgarciae) [reply MaxText/layers/decoders.py:590]: In principle we can, at some point @bvandermoon and I thought it was better to only have it in the constructor but maybe we can have both.

File: MaxText/layers/linears.py (hunk 2 above)
Comment (cgarciae) [reply MaxText/layers/linears.py]: good catch!

File: MaxText/layers/models.py (hunk 8 above)
Comment (cgarciae) [reply MaxText/layers/models.py:318]: I think its always been like this

File: MaxText/tests/model_test.py (hunk 9 above)
Comment (cgarciae) [reply MaxText/tests/model_test.py:170]: this one is better because it gives you more info on the error message

File: MaxText/train_utils.py (hunk 10 above)
Comment (cgarciae) [reply MaxText/train_utils.py:55]: this is helping the type system know that if you pass `rngs` the return is of type `Transformer` (NNX) else its Linen.

**Review by A9isha — APPROVED** [review]
Thank you Cristian - hopefully the tests go green soon
